File: src/java/test/src/androidTest/java/com/tencent/wcdbtest/database/TraceTest.java
// Created by chenqiuwen on 2023/5/10.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdbtest.database;

import org.jetbrains.annotations.NotNull;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.tencent.wcdb.base.Value;
import com.tencent.wcdb.base.WCDBException;
import com.tencent.wcdb.core.Database;
import com.tencent.wcdb.winq.Order;
import com.tencent.wcdb.winq.Pragma;
import com.tencent.wcdb.winq.StatementCreateIndex;
import com.tencent.wcdb.winq.StatementPragma;
import com.tencent.wcdb.winq.StatementSelect;
import com.tencent.wcdbtest.base.FileTool;
import com.tencent.wcdbtest.base.TestObject;
import com.tencent.wcdbtest.base.WrappedValue;
import com.tencent.wcdbtest.base.DBTestObject;
import com.tencent.wcdbtest.base.RandomTool;
import com.tencent.wcdbtest.base.TableTestCase;

import static org.junit.Assert.*;

import android.os.Process;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

@RunWith(AndroidJUnit4.class)
public class TraceTest extends TableTestCase {
    @Test
    public void testTraceSQL() throws WCDBException {
        final StatementPragma statement = new StatementPragma().pragma(Pragma.userVersion);

        final WrappedValue tested = new WrappedValue();
        database.traceSQL(new Database.SQLTracer() {
            @Override
            public void onTrace(long tag, @NotNull String path, long handleId, @NotNull String sql, @NotNull String info) {
                assertEquals(tag, database.getTag());
                assertEquals(path, database.getPath());
                if (sql.equals(statement.getDescription())) {
                    tested.boolValue = true;
                }
            }
        });
        database.execute(statement);
        assertTrue(tested.boolValue);
        database.traceSQL(null);
    }

    @Test
    public void testGlobalTraceSQL() throws WCDBException {
        final StatementPragma statement = new StatementPragma().pragma(Pragma.userVersion);

        final WrappedValue tested = new WrappedValue();
        Database.globalTraceSQL(null);
        Database.globalTraceSQL(new Database.SQLTracer() {
            @Override
            public void onTrace(long tag, @NotNull String path, long handleId, @NotNull String sql, @NotNull String info) {
                if (!database.getPath().equals(path)) {
                    return;
                }
                assertEquals(tag, database.getTag());
                if (sql.equals(statement.getDescription())) {
                    tested.boolValue = true;
                }
            }
        });
        database.execute(statement);
        assertTrue(tested.boolValue);
        Database.globalTraceSQL(null);
    }

    @Test
    public void testTracePerformance() throws WCDBException {
        database.removeFiles();
        createTable();
        database.tracePerformance(null);
        ArrayList<TestObject> objects = new ArrayList<>();
        for (int i = 0; i < 1000; i++) {
            TestObject object = new TestObject();
            object.content = RandomTool.string(4096);
            objects.add(object);
        }
        WrappedValue testCount = new WrappedValue();
        database.tracePerformance(new Database.PerformanceTracer() {
            @Override
            public void onTrace(long tag, @NotNull String path, long handleId, @NotNull String sql, @NotNull Database.PerformanceInfo info) {
                assertEquals(tag, database.getTag());
                assertEquals(path, database.getPath());
                assertNotNull(info);
                if (sql.startsWith("COMMIT")) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertTrue(info.tablePageWriteCount > 0);
                    assertEquals(0, info.indexPageWriteCount);
                    assertTrue(info.overflowPageWriteCount > 0);
                    assertEquals(0, info.tablePageReadCount);
                    assertEquals(0, info.indexPageReadCount);
                    assertEquals(0, info.overflowPageReadCount);
                    testCount.intValue++;
                } else if (sql.startsWith("CREATE INDEX")) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertEquals(1, info.tablePageWriteCount);
                    assertTrue(info.indexPageWriteCount > 0);
                    assertEquals(info.overflowPageWriteCount, objects.size());
                    assertTrue(info.tablePageReadCount > 0);
                    assertTrue(info.indexPageReadCount >= 0);
                    assertTrue(info.overflowPageReadCount > objects.size() / 2);
                    testCount.intValue++;
                } else if (sql.startsWith("SELECT")) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertEquals(0, info.tablePageWriteCount);
                    assertEquals(0, info.indexPageWriteCount);
                    assertEquals(0, info.overflowPageWriteCount);
                    testCount.intValue++;
                    if (sql.endsWith("ORDER BY content DESC")) {
                        assertEquals(0, info.tablePageReadCount);
                        assertTrue(info.indexPageReadCount > 0);
                        assertEquals(info.overflowPageReadCount, objects.size());
                    } else {
                        assertTrue(info.tablePageReadCount > 0);
                        assertEquals(0, info.indexPageReadCount);
                        assertEquals(info.overflowPageReadCount, objects.size());
                    }
                }
            }
        });
        database.insertObjects(objects, DBTestObject.allFields(), tableName);
        database.execute(new StatementCreateIndex().createIndex("testIndex").on(tableName).indexedBy(DBTestObject.content));
        assertEquals(database.getAllObjects(DBTestObject.allFields(), tableName).size(), objects.size());
        assertEquals(database.getAllObjects(DBTestObject.allFields(), tableName, DBTestObject.content.order(Order.Desc)).size(), objects.size());
        assertEquals(testCount.intValue, 4);
        database.tracePerformance(null);
    }

    @Test
    public void testGlobalTracePerformance() throws WCDBException {
        database.removeFiles();
        ArrayList<TestObject> objects = new ArrayList<>();
        for (int i = 0; i < 1000; i++) {
            TestObject object = new TestObject();
            object.content = RandomTool.string(4096);
            objects.add(object);
        }
        WrappedValue testCount = new WrappedValue();
        WrappedValue lastSQLIsInsert = new WrappedValue();
        Database.globalTracePerformance(null);
        Database.globalTracePerformance(new Database.PerformanceTracer() {
            @Override
            public void onTrace(long tag, @NotNull String path, long handleId, @NotNull String sql, @NotNull Database.PerformanceInfo info) {
                if (!database.getPath().equals(path)) {
                    return;
                }
                assertEquals(tag, database.getTag());
                if (sql.startsWith("COMMIT") && lastSQLIsInsert.boolValue) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertTrue(info.tablePageWriteCount > 0);
                    assertEquals(0, info.indexPageWriteCount);
                    assertTrue(info.overflowPageWriteCount > 0);
                    assertEquals(0, info.tablePageReadCount);
                    assertEquals(0, info.indexPageReadCount);
                    assertEquals(0, info.overflowPageReadCount);
                    testCount.intValue++;
                } else if (sql.startsWith("CREATE INDEX")) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertEquals(1, info.tablePageWriteCount);
                    assertTrue(info.indexPageWriteCount > 0);
                    assertEquals(info.overflowPageWriteCount, objects.size());
                    assertTrue(info.tablePageReadCount > 0);
                    assertTrue(info.indexPageReadCount >= 0);
                    assertTrue(info.overflowPageReadCount > objects.size() / 2);
                    testCount.intValue++;
                } else if (sql.startsWith("SELECT")) {
                    assertTrue(info.costInNanoseconds > 0);
                    assertEquals(0, info.tablePageWriteCount);
                    assertEquals(0, info.indexPageWriteCount);
                    assertEquals(0, info.overflowPageWriteCount);
                    testCount.intValue++;
                    if (sql.endsWith("ORDER BY content DESC")) {
                        assertEquals(0, info.tablePageReadCount);
                        assertTrue(info.indexPageReadCount > 0);
                        assertEquals(info.overflowPageReadCount, objects.size());
                    } else {
                        assertTrue(info.tablePageReadCount > 0);
                        assertEquals(0, info.indexPageReadCount);
                        assertEquals(info.overflowPageReadCount, objects.size());
                    }
                }
                lastSQLIsInsert.boolValue = sql.startsWith("INSERT");
            }
        });
        createTable();
        database.insertObjects(objects, DBTestObject.allFields(), tableName);
        database.execute(new StatementCreateIndex().createIndex("testIndex").on(tableName).indexedBy(DBTestObject.content));
        assertEquals(database.getAllObjects(DBTestObject.allFields(), tableName).size(), objects.size());
        assertEquals(database.getAllObjects(DBTestObject.allFields(), tableName, DBTestObject.content.order(Order.Desc)).size(), objects.size());
        assertEquals(testCount.intValue, 4);
        Database.globalTracePerformance(null);
    }

    @Test
    public void testTraceError() {
        final WrappedValue tested = new WrappedValue();
        database.traceException(new Database.ExceptionTracer() {
            @Override
            public void onTrace(@NotNull WCDBException exception) {
                if (exception.level == WCDBException.Level.Error &&
                        exception.path().equals(FileTool.getRealPath(path)) &&
                        exception.tag() == database.getTag() &&
                        exception.code == WCDBException.Code.Error &&
                        exception.sql().equals("SELECT 1 FROM dummy") &&
                        exception.message().equals("no such table: dummy")) {
                    tested.boolValue = true;
                }
            }
        });
        assertTrue(database.canOpen());
        try {
            database.execute(new StatementSelect().select("1").from("dummy"));
        } catch (WCDBException e) {
            assertTrue(tested.boolValue);
        } finally {
            assertTrue(tested.boolValue);
        }
    }

    @Test
    public void testGlobalTraceError() {
        final WrappedValue tested = new WrappedValue();
        Database.globalTraceException(null);
        Database.globalTraceException(new Database.ExceptionTracer() {
            @Override
            public void onTrace(@NotNull WCDBException exception) {
                if (exception.level == WCDBException.Level.Error &&
                        exception.path().equals(FileTool.getRealPath(path)) &&
                        exception.tag() == database.getTag() &&
                        exception.code == WCDBException.Code.Error &&
                        exception.sql().equals("SELECT 1 FROM dummy") &&
                        exception.message().equals("no such table: dummy")) {
                    tested.boolValue = true;
                }
            }
        });
        assertTrue(database.canOpen());
        try {
            database.execute(new StatementSelect().select("1").from("dummy"));
        } catch (WCDBException e) {
            assertTrue(tested.boolValue);
        } finally {
            assertTrue(tested.boolValue);
        }
    }

    @Test
    public void testTraceDBOperation() throws WCDBException {
        final WrappedValue tag = new WrappedValue();
        final WrappedValue path = new WrappedValue();
        final WrappedValue openHandleCount = new WrappedValue();
        final WrappedValue tableCount = new WrappedValue();
        final WrappedValue indexCount = new WrappedValue();
        Database.globalTraceDatabaseOperation(new Database.OperationTracer() {
            @Override
            public void onTrace(@NotNull Database database, Database.Operation operation, @NotNull HashMap<String, Value> infos) {
                switch (operation) {
                    case Create:
                        path.stringValue = database.getPath();
                        break;
                    case SetTag:
                        tag.intValue = database.getTag();
                        break;
                    case OpenHandle:
                        openHandleCount.intValue++;
                        assertEquals(infos.get(Database.OperationInfoKeyHandleCount).getLong(), 1);
                        assertTrue(infos.get(Database.OperationInfoKeyOpenTime).getLong() > 0);
                        assertTrue(infos.get(Database.OperationInfoKeyOpenCPUTime).getLong() > 0);
                        assertTrue(infos.get(Database.OperationInfoKeySchemaUsage).getLong() > 0);
                        assertEquals(infos.get(Database.OperationInfoKeyTriggerCount).getLong(), 0);
                        tableCount.intValue = infos.get(Database.OperationInfoKeyTableCount).getLong();
                        indexCount.intValue = infos.get(Database.OperationInfoKeyIndexCount).getLong();
                        break;
                }
            }
        });
        Database newDatabase = new Database(currentDirectory + File.separator + "testDatabase2");
        newDatabase.setTag(10000);
        newDatabase.createTable(tableName, DBTestObject.INSTANCE);
        newDatabase.execute(new StatementCreateIndex().createIndex("testIndex").on(tableName).indexedBy(DBTestObject.content));

        assertEquals(tag.intValue, newDatabase.getTag());
        assertEquals(path.stringValue, newDatabase.getPath());
        assertEquals(openHandleCount.intValue, 1);

        newDatabase.close();

        newDatabase.insertObjects(RandomTool.autoIncrementTestCaseObjects(10), DBTestObject.allFields(), tableName);

        assertEquals(tableCount.intValue, 4);
        assertEquals(indexCount.intValue, 1);
        assertEquals(openHandleCount.intValue, 2);
        Database.globalTraceDatabaseOperation(null);
    }

    @Test
    public void testGlobalTraceBusy() throws InterruptedException, WCDBException {
        WrappedValue testTid = new WrappedValue();
        Database.globalTraceBusy(new Database.BusyTracer() {
            @Override
            public void onTrace(long tag, @NotNull String path, long tid, @NotNull String sql) {
                assertEquals(tag, database.getTag());
                assertEquals(path, database.getPath());
                assertEquals(sql, "INSERT INTO testTable(id, content) VALUES(?1, ?2)");
                testTid.intValue = tid;
            }
        }, 0.1);
        database.createTable(tableName, DBTestObject.INSTANCE);
        List<TestObject> objects = RandomTool.autoIncrementTestCaseObjects(100000);
        WrappedValue dispatchTid = new WrappedValue();
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                dispatchTid.intValue = Process.myTid();
                database.insertObjects(objects, DBTestObject.allFields(), tableName);
            }
        });
        thread.start();
        sleep(100);
        database.insertObject(RandomTool.testObjectWithId(100001), DBTestObject.allFields(), tableName);
        thread.join();
        assertTrue(testTid.intValue != 0 && testTid.intValue == dispatchTid.intValue);
    }

}

File: src/rust/examples/tests/database/trace_test.rs
use crate::base::random_tool::RandomTool;
use crate::base::table_test_case::TableTestCase;
use crate::base::test_object::{DbTestObject, TestObject};
use crate::base::wrapped_value::WrappedValue;
use std::sync::{Arc, Mutex};
use wcdb::base::wcdb_exception::{ExceptionCode, ExceptionLevel, WCDBException};
use wcdb::core::database::{
    Database, PerformanceInfo, TraceExceptionCallback, TracePerformanceCallback, TraceSqlCallback,
};
use wcdb::core::handle_operation::HandleOperationTrait;
use wcdb::core::handle_orm_operation::HandleORMOperationTrait;
use wcdb::winq::identifier::IdentifierTrait;
use wcdb::winq::ordering_term::Order;
use wcdb::winq::pragma::Pragma;
use wcdb::winq::statement_create_index::StatementCreateIndex;
use wcdb::winq::statement_pragma::StatementPragma;
use wcdb::winq::statement_select::StatementSelect;

static TABLE_NAME: &'static str = "trace_test_case";
pub struct TraceTest {
    table_test_case: TableTestCase,
}

impl TraceTest {
    pub fn new() -> Self {
        TraceTest {
            table_test_case: TableTestCase::new_with_table_name(TABLE_NAME),
        }
    }
}

impl TraceTest {
    pub fn test_trace_sql(&self) {
        let statement = StatementPragma::new();
        let statement = statement.pragma(Pragma::user_version());
        let desc = statement.get_description();
        let mut tested = Arc::new(Mutex::new(WrappedValue::new()));
        let tested_clone = tested.clone();

        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        let db_tag = database.get_tag();
        let db_path = database.get_path();

        database
            .trace_sql(Some(
                move |tag: i64, path: String, handle_id: i64, sql: String, info: String| {
                    assert_eq!(tag, db_tag);
                    assert_eq!(path, db_path);
                    if sql.eq(desc.as_str()) {
                        tested.lock().unwrap().bool_value = true;
                    }
                },
            ))
            .unwrap();

        database.execute(statement).unwrap();
        assert!(tested_clone.lock().unwrap().bool_value);
        database.trace_sql::<TraceSqlCallback>(None).unwrap();
    }

    pub fn test_global_trace_sql(&self) {
        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        database.remove_files().unwrap();

        let statement = StatementPragma::new();
        let statement = statement.pragma(Pragma::user_version());
        let desc = statement.get_description();
        let mut tested = Arc::new(Mutex::new(WrappedValue::new()));
        let tested_clone = tested.clone();
        let db_tag = database.get_tag();
        let db_path = database.get_path();

        Database::global_trace_sql::<TraceSqlCallback>(None).unwrap();
        Database::global_trace_sql(Some(
            move |tag: i64, path: String, handle_id: i64, sql: String, info: String| {
                if !path.eq(db_path.as_str()) {
                    return;
                }
                assert_eq!(tag, db_tag);
                if sql.eq(desc.as_str()) {
                    tested.lock().unwrap().bool_value = true;
                }
            },
        ))
        .unwrap();
        database.execute(statement).unwrap();
        assert!(tested_clone.lock().unwrap().bool_value);
        Database::global_trace_sql::<TraceSqlCallback>(None).unwrap();
    }

    pub fn test_trace_performance(&self) {
        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        database.remove_files().unwrap();

        self.table_test_case.create_table().unwrap();
        let mut objects = vec![];
        let obj_size = 1000;
        for _ in 0..obj_size {
            let content = RandomTool::string_by_length(4096);
            let object = TestObject::new(content);
            objects.push(object);
        }

        let mut test_count = Arc::new(Mutex::new(WrappedValue::new()));
        let test_count_clone = test_count.clone();
        let db_tag = database.get_tag();
        let db_path = database.get_path();

        database
            .trace_performance::<TracePerformanceCallback>(None)
            .unwrap();
        database
            .trace_performance(Some(
                move |tag: i64,
                      path: String,
                      handle_id: i64,
                      sql: String,
                      info: PerformanceInfo| {
                    assert_eq!(tag, db_tag);
                    assert_eq!(path, db_path);
                    if sql.starts_with("COMMIT") {
                        assert!(info.cost_in_nanoseconds > 0);
                        assert!(info.table_page_write_count > 0);
                        assert_eq!(0, info.index_page_write_count);
                        assert!(info.overflow_page_write_count > 0);
                        assert_eq!(0, info.table_page_read_count);
                        assert_eq!(0, info.index_page_read_count);
                        assert_eq!(0, info.overflow_page_read_count);
                        test_count.lock().unwrap().int_value += 1;
                    } else if sql.starts_with("CREATE INDEX") {
                        assert!(info.cost_in_nanoseconds > 0);
                        assert_eq!(1, info.table_page_write_count);
                        assert!(info.index_page_write_count > 0);
                        assert_eq!(info.overflow_page_write_count, obj_size);
                        assert!(info.table_page_read_count > 0);
                        assert!(info.index_page_read_count >= 0);
                        assert!(info.overflow_page_read_count > obj_size / 2);
                        test_count.lock().unwrap().int_value += 1;
                    } else if sql.starts_with("SELECT") {
                        assert!(info.cost_in_nanoseconds > 0);
                        assert_eq!(0, info.table_page_write_count);
                        assert_eq!(0, info.index_page_write_count);
                        assert_eq!(0, info.overflow_page_write_count);
                        test_count.lock().unwrap().int_value += 1;
                        if sql.ends_with("ORDER BY content DESC") {
                            assert_eq!(0, info.table_page_read_count);
                            assert!(info.index_page_read_count > 0);
                            assert_eq!(info.overflow_page_read_count, obj_size);
                        } else {
                            assert!(info.table_page_read_count > 0);
                            assert_eq!(0, info.index_page_read_count);
                            assert_eq!(info.overflow_page_read_count, obj_size);
                        }
                    }
                },
            ))
            .unwrap();

        database
            .insert_objects(objects, DbTestObject::all_fields(), TABLE_NAME)
            .unwrap();

        database
            .execute(
                StatementCreateIndex::new()
                    .create_index("testIndex")
                    .on(TABLE_NAME)
                    .indexed_by(vec![DbTestObject::content().get_column()]),
            )
            .unwrap();

        assert_eq!(
            database
                .get_all_objects(
                    DbTestObject::all_fields(),
                    TABLE_NAME,
                    None,
                    None,
                    None,
                    None
                )
                .unwrap()
                .len(),
            obj_size as usize
        );

        assert_eq!(
            database
                .get_all_objects(
                    DbTestObject::all_fields(),
                    TABLE_NAME,
                    None,
                    Some(DbTestObject::content().get_column().order(Order::Desc)),
                    None,
                    None
                )
                .unwrap()
                .len(),
            obj_size as usize
        );

        assert_eq!(test_count_clone.lock().unwrap().int_value, 4);
        database
            .trace_performance::<TracePerformanceCallback>(None)
            .unwrap();
    }

    pub fn test_global_trace_performance(&self) {
        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        database.remove_files().unwrap();
        let mut objects = vec![];
        let obj_size = 1000;
        for _ in 0..obj_size {
            let content = RandomTool::string_by_length(4096);
            let object = TestObject::new(content);
            objects.push(object);
        }

        let mut test_count = Arc::new(Mutex::new(WrappedValue::new()));
        let test_count_clone = test_count.clone();
        let mut last_sql_is_insert = Arc::new(Mutex::new(WrappedValue::new()));

        let db_tag = database.get_tag();
        let db_path = database.get_path();

        Database::global_trace_performance::<TracePerformanceCallback>(None).unwrap();
        Database::global_trace_performance(Some(
            move |tag: i64, path: String, handle_id: i64, sql: String, info: PerformanceInfo| {
                if !path.eq(&db_path) {
                    return;
                }
                assert_eq!(tag, db_tag);
                if sql.starts_with("COMMIT") && last_sql_is_insert.lock().unwrap().bool_value {
                    assert!(info.cost_in_nanoseconds > 0);
                    assert!(info.table_page_write_count >= 0);
                    assert!(info.index_page_write_count >= 0);
                    assert!(info.overflow_page_write_count >= 0);
                    assert_eq!(0, info.table_page_read_count);
                    assert_eq!(0, info.index_page_read_count);
                    assert_eq!(0, info.overflow_page_read_count);
                    test_count.lock().unwrap().int_value += 1;
                } else if sql.starts_with("CREATE INDEX") {
                    assert!(info.cost_in_nanoseconds > 0);
                    assert_eq!(1, info.table_page_write_count);
                    assert!(info.index_page_write_count > 0);
                    assert_eq!(info.overflow_page_write_count, obj_size);
                    assert!(info.table_page_read_count > 0);
                    assert!(info.index_page_read_count >= 0);
                    assert!(info.overflow_page_read_count > obj_size / 2);
                    test_count.lock().unwrap().int_value += 1;
                } else if sql.starts_with("SELECT") {
                    assert!(info.cost_in_nanoseconds > 0);
                    assert_eq!(0, info.table_page_write_count);
                    assert_eq!(0, info.index_page_write_count);
                    assert_eq!(0, info.overflow_page_write_count);
                    test_count.lock().unwrap().int_value += 1;
                    if sql.ends_with("ORDER BY content DESC") {
                        assert_eq!(0, info.table_page_read_count);
                        assert!(info.index_page_read_count > 0);
                        assert_eq!(info.overflow_page_read_count, obj_size);
                    } else {
                        assert!(info.table_page_read_count > 0);
                        assert_eq!(0, info.index_page_read_count);
                        assert_eq!(info.overflow_page_read_count, obj_size);
                    }
                }
                last_sql_is_insert.lock().unwrap().bool_value = sql.starts_with("INSERT");
            },
        ))
        .unwrap();

        self.table_test_case.create_table().unwrap();
        database
            .insert_objects(objects, DbTestObject::all_fields(), TABLE_NAME)
            .unwrap();

        database
            .execute(
                StatementCreateIndex::new()
                    .create_index("testIndex")
                    .on(TABLE_NAME)
                    .indexed_by(vec![DbTestObject::content().get_column()]),
            )
            .unwrap();

        assert_eq!(
            database
                .get_all_objects(
                    DbTestObject::all_fields(),
                    TABLE_NAME,
                    None,
                    None,
                    None,
                    None
                )
                .unwrap()
                .len(),
            obj_size as usize
        );

        assert_eq!(
            database
                .get_all_objects(
                    DbTestObject::all_fields(),
                    TABLE_NAME,
                    None,
                    Some(DbTestObject::content().get_column().order(Order::Desc)),
                    None,
                    None
                )
                .unwrap()
                .len(),
            obj_size as usize
        );

        assert_eq!(test_count_clone.lock().unwrap().int_value, 4);
        Database::global_trace_performance::<TracePerformanceCallback>(None).unwrap();
    }

    pub fn test_trace_error(&self) {
        let mut tested = Arc::new(Mutex::new(WrappedValue::new()));
        let tested_clone = tested.clone();
        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        let db_tag = database.get_tag();
        let db_path = database.get_path();
        database
            .trace_exception(Some(move |exception: WCDBException| {
                if exception.level == ExceptionLevel::Error
                    && exception.path().eq(db_path.as_str())
                    && exception.tag() == db_tag
                    && exception.code == ExceptionCode::Error
                    && exception.sql().eq("SELECT 1 FROM dummy")
                    && exception.message().eq("no such table: dummy")
                {
                    tested.lock().unwrap().bool_value = true;
                }
            }))
            .unwrap();

        assert!(database.can_open());
        let ret = database.execute(
            StatementSelect::new()
                .select(vec!["1"])
                .from("dummy" /* table_subquery_obj_vec */),
        );
        match ret {
            Ok(_) => {
                assert!(tested_clone.lock().unwrap().bool_value);
            }
            Err(_) => {
                assert!(tested_clone.lock().unwrap().bool_value);
            }
        }
        database
            .trace_exception::<TraceExceptionCallback>(None)
            .unwrap();
    }

    pub fn test_global_trace_error(&self) {
        let mut tested = Arc::new(Mutex::new(WrappedValue::new()));
        let tested_clone = tested.clone();
        let database_arc = self.table_test_case.get_database().clone();
        let database = database_arc.read().unwrap();
        let db_tag = database.get_tag();
        let db_path = database.get_path();

        Database::global_trace_exception::<TraceExceptionCallback>(None).unwrap();

        Database::global_trace_exception(Some(move |exception: WCDBException| {
            if exception.level == ExceptionLevel::Error
                && exception.path().eq(db_path.as_str())
                && exception.tag() == db_tag
                && exception.code == ExceptionCode::Error
                && exception.sql().eq("SELECT 1 FROM dummy")
                && exception.message().eq("no such table: dummy")
            {
                tested.lock().unwrap().bool_value = true;
            }
        }))
        .unwrap();

        assert!(database.can_open());
        let ret = database.execute(
            StatementSelect::new()
                .select(vec!["1".to_string()])
                .from("dummy"),
        );
        match ret {
            Ok(_) => {
                assert!(tested_clone.lock().unwrap().bool_value);
            }
            Err(_) => {
                assert!(tested_clone.lock().unwrap().bool_value);
            }
        }
        Database::global_trace_exception::<TraceExceptionCallback>(None).unwrap();
    }
}

#[cfg(test)]
pub mod trace_test {
    use crate::database::trace_test::TraceTest;

    // #[test]
    // todo qixinbing 单独运行正常，但是集成测试会失败
    pub fn test_trace_sql() {
        let trace_test = TraceTest::new();
        trace_test.test_trace_sql();
    }

    #[test]
    pub fn test_global_trace_sql() {
        let trace_test = TraceTest::new();
        trace_test.test_global_trace_sql();
    }

    // #[test]
    // todo qixinbing 单独运行正常，但是集成测试会失败
    pub fn test_trace_performance() {
        let trace_test = TraceTest::new();
        trace_test.test_trace_performance();
    }

    // #[test]
    // todo qixinbing 本地单独、集测运行正常，但是 ci 集成测试会失败
    pub fn test_global_trace_performance() {
        let trace_test = TraceTest::new();
        trace_test.test_global_trace_performance();
    }

    #[test]
    pub fn test_trace_error() {
        let trace_test = TraceTest::new();
        trace_test.test_trace_error();
    }

    #[test]
    pub fn test_global_trace_error() {
        let trace_test = TraceTest::new();
        trace_test.test_global_trace_error();
    }
}
